### DISCUSSION TITLE
Better dynamic support

### DISCUSSION THREAD

**Discussion by bbarry** [dotnet/csharplang#284] [General]
Original: [dotnet/roslyn/#3012](https://github.com/dotnet/roslyn/issues/3012)

The generic type syntax could be used with `dynamic` as `dynamic<T>` to give tooling support about members and operators related to the usage of the variable.

This tooling could include:

1. member lookup on the instance
2. warnings/errors where it can be known that a member doesn't exist or an operation is invalid

**Comment by TonyValenti** [comment]
I've long thought that "dynamic" should be a variable modifier that would do something similar to what you indicated.

For example:

dynamic foo = SomeMethod(...); ///We can do this today
dynamic Person foo = CreatePerson(...);  // This should let me Dot-access person members and DLR-access non-person members.
dynamic var foo = CreateSuperMan(...); // This should allow me to dot-access any publicly accessible members from the SuperMan class and DLR-access other members.

**Comment by bbarry** [comment]
As a related thought: [dotnet/roslyn/#5306](https://github.com/dotnet/roslyn/issues/5306) is also looking for some compile time enforcement of `dynamic` types.

**Comment by TonyValenti** [comment]
Hi All,
I was also thinking about this in reference to #228 (Intersection Types).

Perhaps we should just allow listing multiple types to create intersection types.  For example, here are a bunch of declarations:
````
var Item = ...
dynamic var Item = ...
IStudent ITeacher = ...
dynamic IStudent = ...
dynamic IStudent ITeacher IEmployable IActionable = ...
````

**Comment by Thaina** [comment]
@TonyValenti There are proposal about union and intersect type #344 #399 which is the same as typescript approach

Also I prefer `dynamic<>` syntax I have seen in #969

```C#
using CompoundType = (TU1 | TU2) & TI1;

dynamic<CompoundType> obj = ...;
```
